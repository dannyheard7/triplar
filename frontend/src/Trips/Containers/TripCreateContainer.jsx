import React from "react";
import axios from "axios";

import TripForm from "Trips/Components/TripForm";

export default class TripCreate extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            trip: {},
            errors: []
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.create = this.create.bind(this);
    }

    handleChange(event) {
        let trip = this.state.trip;
        trip[event.target.name] = event.target.value;

        this.setState({trip: trip});
    }

    handleSubmit(e) {
        e.preventDefault();
        this.create()
    }

    create () {
        axios.post("/trips/", this.state.trip).then(response => {
            if (response.status === 201) {
                this.setState({trip: {}, errors: []});
                this.props.onTripCreate(response.data);
            } else if (response.status === 400) {
                this.setState({errors: response.data})
            }
        });
    }

    render () {
        return (
            <div className="card">
                <TripForm onChange={this.handleChange} onSubmit={this.handleSubmit} submitLabel="Create"
                           errors={this.state.errors} />
            </div>
        )
    }
}