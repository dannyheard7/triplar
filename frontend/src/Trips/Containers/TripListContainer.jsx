import React from "react";
import api from "utils/api.js";

import TripList from "Trips/Components/TripList";

export default class TripListContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            trips: []
        };
    }

    componentDidMount() {
        this.getTrips();
    }

    componentWillReceiveProps(props) {
        if(props.update) {
            this.getTrips();
        }
    }

    getTrips() {
        api.getTrips().then(res => {
            this.setState({trips: res.data});
        });
    }

    render() {
        return (<TripList trips={this.state.trips}/>);
    }
}