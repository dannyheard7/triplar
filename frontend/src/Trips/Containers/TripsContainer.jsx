import React from "react";

import TripListContainer from "Trips/Containers/TripListContainer";
import TripCreateContainer from "Trips/Containers/TripCreateContainer";

export default class TripsContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showTripCreate: false,
            updateTripsList: false
        };

        this.onClick = this.onClick.bind(this);
        this.onTripCreate = this.onTripCreate.bind(this);
    }

    onClick() {
        this.setState({showTripCreate: !this.state.showTripCreate, updateTripsList: false});
    }

    onTripCreate() {
        this.setState({
            showTripCreate: false,
            updateTripsList: true,
        });
    }

    render() {
        return (
            <div>
                <TripListContainer update={this.state.updateTripsList}/>
                <div className="row">
                    <div className="col md-2">
                        <button className="btn btn-primary" onClick={this.onClick}>New Trip</button>
                    </div>
                </div>
                {this.state.showTripCreate && <TripCreateContainer onTripCreate={this.onTripCreate} />}
            </div>
        )
    }
}