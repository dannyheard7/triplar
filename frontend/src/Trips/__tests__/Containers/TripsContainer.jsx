import React from "react";
import {shallow} from "enzyme";

import TripListContainer from "Trips/Containers/TripListContainer";
import TripCreateContainer from "Trips/Containers/TripCreateContainer";
import TripsContainer from "Trips/Containers/TripsContainer";
import ShallowRenderer from "react-test-renderer/shallow";

const renderer = new ShallowRenderer();

describe('<TripsContainer />', () => {
    let overview = null;
    const event = {
        preventDefault: () => {},
        target: {
            value: "",
            name: "",
        },
    };

    beforeEach(function () {
        overview = shallow(<TripsContainer />);
    });

    test('renders a <TripListContainer /> component', () => {
        const result = renderer.render(<TripListContainer  />);
        expect(result).toMatchSnapshot();
    });

    describe('trip create visible', () => {
        beforeEach(function () {
            overview.setState({showTripCreate: true});
        });

        it('renders <TripCreateContainer /> component when tripCreateVisible is true', () => {
            expect(overview.find(TripCreateContainer).length).toEqual(1);
        });

        it('updates state.tripCreateVisible when create button clicked', () => {
            overview.find('.btn').simulate('click');
            expect(overview.state('showTripCreate')).toEqual(false);
        });

        it('does not render <TripCreateContainer /> after a trip has been created', () => {
            expect(overview.find(TripCreateContainer).length).toEqual(1);
            overview.find(TripCreateContainer).prop('onTripCreate')(event);
            overview.update();

            expect(overview.state('showTripCreate')).toEqual(false);
            expect(overview.find(TripCreateContainer).length).toEqual(0);
        });

        it('The state contains the new trip from <TripCreateContainer /> after a trip has been created', () => {
            let trip = jest.fn();
            overview.find(TripCreateContainer).prop('onTripCreate')(trip);

            expect(overview.state('createdTrip')).toEqual(trip);
        });
    });

    it('createdTrip state change updates TripListContainer Prop', function () {
        const container = shallow(<TripsContainer />);
        expect(container.find(TripListContainer).props().createdTrip).toEqual(container.state('createdTrip'));

        let createdTrip = jest.fn();
        container.setState({createdTrip: createdTrip});
        expect(container.find(TripListContainer).props().createdTrip).toEqual(createdTrip);
    });

    test.skip('Once trip has been created, all state of the new trip is cleared', () => {

    })
});