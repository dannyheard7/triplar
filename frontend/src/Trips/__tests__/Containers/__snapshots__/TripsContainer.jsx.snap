// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripsContainer /> renders a <TripListContainer /> component 1`] = `
<TripList
  trips={Array []}
/>
`;
